components/supervisor/pkg/supervisor/services.go

```diff
@@ -1033,3 +1035,149 @@ func (s *portService) RetryAutoExpose(ctx context.Context, req *api.RetryAutoExp
+	return api.RegisterPortServiceHandlerFromEndpoint(context.Background(), mux, grpcEndpoint, []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())})
+}
+
+// ListenToOutput listens to the output of a task. It streams the output from the task's file and ends when the task's state changes to done.
+func (s *taskService) ListenToOutput(req *api.ListenToOutputRequest, srv api.TaskService_ListenToOutputServer) error {
+	fileLocation := logs.PrebuildLogFileName(logs.TerminalStoreLocation, req.TaskId)
+	if _, err := os.Stat(fileLocation); os.IsNotExist(err) {
+		return status.Error(codes.NotFound, "task not found")
+	}
+
+	file, err := os.Open(fileLocation)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer file.Close()
+
+	buf := make([]byte, 4096)
+
+	// If the task is already done, we can just send all the logs and return without subscribing to task updates
+	taskStatuses := s.tasksManager.Status()
+	taskClosed := false
+	for _, taskStatus := range taskStatuses {
+		if taskStatus.Id == req.TaskId && taskStatus.State == api.TaskState_closed {
+			taskClosed = true
+			break
+		}
+	}
+
+	if taskClosed {
+		buf := make([]byte, 4096)
+		for {
+			n, err := file.Read(buf)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return status.Error(codes.Internal, err.Error())
+			}
+
+			if err := srv.Send(&api.ListenToOutputResponse{Data: buf[:n]}); err != nil {
+				return status.Error(codes.Internal, err.Error())
+			}
+		}
+
+		return nil
+	}
+
+	done := make(chan struct{})
+
+	sub := s.tasksManager.Subscribe()
+	if sub == nil {
+		log.Warn("potentially leaking subscription to tasks status: too many subscriptions")
+		return status.Error(codes.ResourceExhausted, "too many subscriptions")
+	}
+	defer sub.Close()
+
+	go func() {
+		defer close(done)
+		for {
+			select {
+			case <-srv.Context().Done():
+				return
+			case updates := <-sub.Updates():
+				if updates == nil {
+					return
+				}
+				for _, update := range updates {
+					if update == nil {
+						continue
+					}
+					if update.Id == req.TaskId && update.State == api.TaskState_closed {
+						return
+					}
+				}
+			}
+		}
+	}()
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer watcher.Close()
+
+	dir, fileName := filepath.Split(fileLocation)
+
+	err = watcher.Add(dir)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	// Send all previous logs before starting to watch for new ones
+	offset := int64(0)
+	for {
+		n, err := file.ReadAt(buf, offset)
+		if n > 0 {
+			offset += int64(n)
+			if err := srv.Send(&api.ListenToOutputResponse{Data: buf[:n]}); err != nil {
+				return status.Error(codes.Internal, err.Error())
+			}
+		}
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	for {
```

Comment:
I feel the three for loops are a bit more complicated then it needs to be, and could probably folded into one. It woulds also decrease the risk for getting ouf of sync.